src/coprocessor_v2/plugin_registry.rs

```diff
@@ -2,28 +2,128 @@
-    loaded_plugins: BTreeMap<String, LoadedPlugin>,
+    inner: Arc<RwLock<PluginRegistryInner>>,
+    /// Active file system watcher for hot-reloading.
+    /// If dropped, the corresponding hot-reloading thread will stop.
+    fs_watcher: Option<notify::RecommendedWatcher>,
 }
 
 #[allow(dead_code)]
 impl PluginRegistry {
     /// Creates a new `PluginRegistry`.
     pub fn new() -> Self {
-        PluginRegistry::default()
+        let registry = Arc::new(RwLock::new(PluginRegistryInner::default()));
+
+        PluginRegistry {
+            inner: registry,
+            fs_watcher: None,
+        }
+    }
+
+    /// Hot-reloads plugins from a given directory.
+    ///
+    /// All plugins that are already present in the directory will be loaded.
+    /// A background thread is spawned to watch file system events. If the library file of a loaded
+    /// plugin is deleted, the corresponding plugin is automatically unloaded; if a new library file
+    /// is placed into the directory, it will be automatically loaded into TiKV's coprocessor plugin
+    /// system.
+    ///
+    /// A file will only be loaded if it has the proper file ending of dynamic link libraries for
+    /// the current platform (`.so` for Linux, `.dylib` for MacOS, `.dll` for Windows).
+    pub fn start_hot_reloading(
+        &mut self,
+        plugin_directory: impl Into<PathBuf>,
+    ) -> notify::Result<()> {
+        let plugin_directory = plugin_directory.into();
+
+        // If this is the first call to `start_hot_reloading()`, create a new file system watcher
+        // and background thread for loading plugins. For later invocations, the same watcher and
+        // thread will be used.
+        if self.fs_watcher.is_none() {
+            let (tx, rx) = mpsc::channel();
+            let fs_watcher = notify::watcher(tx, Duration::from_secs(3)).unwrap();
+
+            let hot_reload_registry = self.inner.clone();
+            thread::spawn(move || {
+                // Simple helper functions for loading/unloading plugins.
+                let maybe_load = |file: &PathBuf| {
+                    if is_library_file(&file) {
+                        // Discard error.
+                        let _ = hot_reload_registry.write().unwrap().load_plugin(file);
+                    }
+                };
+                let unload = |file: &PathBuf| {
+                    hot_reload_registry
+                        .write()
+                        .unwrap()
+                        .unload_plugin_by_path(file)
+                };
+
+                loop {
+                    match rx.recv() {
+                        Ok(DebouncedEvent::Create(file)) => {
+                            maybe_load(&file);
+                        }
+                        Ok(DebouncedEvent::Remove(file)) => {
+                            unload(&file);
+                        }
+                        Ok(DebouncedEvent::Write(file)) => {
+                            unload(&file);
```

Comment:
Will modifying content of an opened executable or dynamic library lead to undefined behavior? Maybe we should warn our users for this case.

Reply:
On Linux, an executable can even be removed without affecting the process. And on Windows, you will be forbidden from modifying it.

Reply:
Removing / recreating won't cause issue. But rewriting would. https://unix.stackexchange.com/questions/404551/overwriting-a-running-executable-or-so

Reply:
hm good point. but does it matter when we fully unload and re-load it after the write is done?

Reply:
We could not determine whether the plugin has *fully* unloaded. There might be some futures running in tokio.

Reply:
hm I see...  good point.
but not sure what we can do about it? the only thing we could do is log a warning when a plugin is overwritten on the file system. or do you have any other ideas?

Reply:
I think logging a warning to our user is enough.

(By the way, I came with another possible fault: user moves the plugin outside of directory and rewrites its content, this is something we could not detect. We just do what we can, to only handle the case that plugin is rewritten inside the plugin directory and print a warning by `eprintln!` or `warn!`.)

Reply:
I just tried out what happens when we use `std::fs::rename` where the target is in a different directory. `notify` actually issues a `Remove` instead of a `Rename`, so we will just unload the plugin because it was basically moved outside the plugin directory. I think this behavior is fine. :)

Reply:
@skyzh added a warning message when a plugin is overwritten :) PTAL